src/propagators.jl

```diff
@@ -0,0 +1,208 @@
+
+immutable QuPropagatorState
+    psi::QuVector
+    t
+    t_state
+end
+
+@doc """
+Input Parameters : QuPropagator
+
+Returns the starting, i.e., the initial state of the system
+""" ->
+function Base.start(prob::QuPropagator)
+    t_state = start(prob.tlist)
+    t,t_state = next(prob.tlist,t_state)
+    return QuPropagatorState(prob.init_state,t,t_state)
+end
+
+@doc """
+Input Parameters : QuPropagator and QuPropagator State
+
+Returns the next state by dispatching to particular
+`propagate` method depending on the `numerical` method type.
+""" ->
+function Base.next{QPM<:QuPropagatorMethod}(prob::QuPropagator{QPM}, qustate::QuPropagatorState)
+    op = prob.hamiltonian
+    current_qustate = qustate.psi
+    current_t = qustate.t
+    if done(prob, qustate) == false
+        t,t_state = next(prob.tlist, qustate.t_state)
+        next_qustate = propagate(prob.method, op, t, current_t, current_qustate)
+        return (next_qustate, t), QuPropagatorState(next_qustate, t, t_state)
+    else
+        return QuPropagatorState(current_qustate, current_t, qustate.t_state)
+    end
+end
+
+@doc """
+Input Parameters : QuPropagator and QuPropagator State
+
+Returns true if the current state is final state, else false
+""" ->
+Base.done(prob::QuPropagator, qustate::QuPropagatorState) = done(prob.tlist, qustate.t_state)
+
+@doc """
+Euler Method
+
+Step Propagation using the Euler formula.
+""" ->
+immutable QuEuler <: QuPropagatorMethod
+end
+
+@doc """
+Crank Nicolson Method
+
+Step Propagation using the Crank Nicolson formula.
+""" ->
+immutable QuCN <: QuPropagatorMethod
+end
+
+@doc """
+Krylov subspace Method
+Input Parameters :
+`options` : Dictionary to set the basis size with keyword NC.
+
+For ex :
+x = [:NC => basis_size]
+QuKrylov(x)
+Step Propagation using the Krylov subspace iterations.
+""" ->
+immutable QuKrylov <: QuPropagatorMethod
+    options::Dict
+end
+
+@doc """
+Propagates to the next time state
+Input Parameters:
+`prob`             :  Method to be used
+`op`               :  Hamiltonian of the system
+`t`                :  Time corresponding to the t_state
+`current_t`        :  Current time
+`current_qustate`  :  Quantum state corresponding to current time
+""" ->
+function propagate(prob::QuEuler, op, t, current_t, current_qustate)
+    dt = t - current_t
+    return (eye(op)-im*op*dt)*current_qustate
+end
+
+function propagate(prob::QuCN, op, t, current_t, current_qustate)
+    dt = t - current_t
+    uni = eye(op)-im*op*dt/2
+    return \(uni', uni*current_qustate)
+end
+
+function propagate(prob::QuKrylov, op, t, current_t, current_qustate)
+    basis_size = prob.options[:NC]
+    dt = t-current_t
+    N = min(basis_size, length(current_qustate))
+    v = Array(typeof(current_qustate),0)
+    push!(v,zeros(current_qustate))
+    push!(v,current_qustate)
+    w = Array(typeof(current_qustate),0)
```

Comment:
I think you don't need to store the `w` of each iteration. `w` can just hold the vector calculated in the respective step.